File: app/src/main/java/com/vpulse/ftpnext/navigation/NavigationDelete.java
package com.vpulse.ftpnext.navigation;

import android.content.DialogInterface;
import android.os.Handler;
import android.view.View;
import android.widget.CheckBox;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;

import com.vpulse.ftpnext.R;

import org.apache.commons.net.ftp.FTPFile;

import static com.vpulse.ftpnext.navigation.NavigationActivity.NAVIGATION_MESSAGE_DELETE_FINISHED;
import static com.vpulse.ftpnext.navigation.NavigationActivity.NAVIGATION_ORDER_DISMISS_DIALOGS;
import static com.vpulse.ftpnext.navigation.NavigationActivity.NAVIGATION_ORDER_REFRESH_DATA;
import static com.vpulse.ftpnext.navigation.NavigationActivity.NAVIGATION_ORDER_SELECTED_MODE_OFF;

public class NavigationDelete {

    private final static String TAG = "NAVIGATION DELETE";

    private final NavigationActivity mContextActivity;
    private final Handler mHandler;

    private NavigationDelete() throws InstantiationException {
        mContextActivity = null;
        mHandler = null;
        throw new InstantiationException("Constructor not allowed");
    }

    protected NavigationDelete(NavigationActivity iContextActivity, Handler iHandler) {
        mContextActivity = iContextActivity;
        mHandler = iHandler;
    }

    protected void onResume() {

    }

    protected void createDialogDeleteSelection() {
        final FTPFile[] lSelectedFiles = mContextActivity.mCurrentAdapter.getSelection();

        if (lSelectedFiles.length == 0)
            mContextActivity.createDialogError("Select something.").show();
        else {
            mContextActivity.mDeletingInfoDialog = new AlertDialog.Builder(mContextActivity)
                    .setTitle("Deleting :") // TODO : Strings
                    .setMessage("Are you sure to delete the selection ? (" + lSelectedFiles.length + " files)")
                    .setPositiveButton("yes", new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface iDialog, int iWhich) {
                            iDialog.dismiss();
                            deleteFile(lSelectedFiles);
                        }
                    })
                    .setNegativeButton("cancel", null)
                    .show();
        }
    }

    private void deleteFile(FTPFile[] iSelectedFiles) {
        mHandler.sendEmptyMessage(NAVIGATION_ORDER_SELECTED_MODE_OFF);
        mContextActivity.mFTPServices.deleteFiles(iSelectedFiles, mContextActivity.mFTPServices.new OnDeleteListener() {

            ProgressBar mProgressDirectory;
            ProgressBar mProgressSubDirectory;
            TextView mTextViewDirectory;
            TextView mTextViewSubDirectory;

            @Override
            public void onStartDelete() {
                mHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        final AlertDialog.Builder lBuilder = new AlertDialog.Builder(mContextActivity);
                        lBuilder.setTitle("Deleting"); // TODO : strings

                        View lDialogDoubleProgress = View.inflate(mContextActivity, R.layout.dialog_double_progress, null);
                        mProgressDirectory = lDialogDoubleProgress.findViewById(R.id.dialog_double_progress_progress_1);
                        mTextViewDirectory = lDialogDoubleProgress.findViewById(R.id.dialog_double_progress_text_1);
                        mProgressSubDirectory = lDialogDoubleProgress.findViewById(R.id.dialog_double_progress_progress_2);
                        mTextViewSubDirectory = lDialogDoubleProgress.findViewById(R.id.dialog_double_progress_text_2);

                        lBuilder.setView(lDialogDoubleProgress);
                        lBuilder.setCancelable(false);
                        lBuilder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() { // TODO : Strings
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                stopDeleting();
                            }
                        });
                        mContextActivity.mDeletingInfoDialog = lBuilder.create();
                        mContextActivity.mDeletingInfoDialog.show();
                    }
                });
            }

            @Override
            public void onProgressDirectory(final int iDirectoryProgress, final int iTotalDirectories, final String iDirectoryName) {
                mHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        mTextViewDirectory.setText(iDirectoryName);
                        mProgressDirectory.setMax(iTotalDirectories);
                        mProgressDirectory.setProgress(iDirectoryProgress);
                    }
                });
            }

            @Override
            public void onProgressSubDirectory(final int iSubDirectoryProgress, final int iTotalSubDirectories, final String iSubDirectoryName) {
                mHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        mTextViewSubDirectory.setText(iSubDirectoryName);
                        mProgressSubDirectory.setMax(iTotalSubDirectories);
                        mProgressSubDirectory.setProgress(iSubDirectoryProgress);
                    }
                });
            }

            @Override
            public void onRightAccessFail(final FTPFile iFTPFile) {
                super.onRightAccessFail(iFTPFile);
                mHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        final AlertDialog.Builder lBuilder = new AlertDialog.Builder(mContextActivity);
                        lBuilder.setTitle("Permission error"); // TODO : strings
                        final View lDialogDeleteError = View.inflate(mContextActivity, R.layout.dialog_delete_error, null);
                        final CheckBox lCheckBox = lDialogDeleteError.findViewById(R.id.dialog_delete_error_checkbox);
                        final TextView lTextView = lDialogDeleteError.findViewById(R.id.dialog_delete_error_text);

                        lTextView.setText("Remember this choice : "); // TODO : strings
                        lBuilder.setView(lDialogDeleteError);
                        String lMessage =
                                "\nNot enough permissions to delete this " +
                                        (iFTPFile.isDirectory() ? "directory \n" : "file \n") +
                                        iFTPFile.getName();
                        lBuilder.setMessage(lMessage);
                        lBuilder.setCancelable(false);
                        lBuilder.setPositiveButton("Ignore", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface iDialog, int iWhich) {
                                iDialog.dismiss();
                                mContextActivity.mFTPServices.setDeletingByPassRightErrors(lCheckBox.isChecked());
                                mContextActivity.mFTPServices.resumeDeleting();
                            }
                        });

                        lBuilder.setNegativeButton("Stop", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface iDialog, int iWhich) {
                                iDialog.dismiss();
                                stopDeleting();
                            }
                        });
                        mContextActivity.mDeletingErrorDialog = lBuilder.create();
                        mContextActivity.mDeletingErrorDialog.show();
                    }
                });
            }

            @Override
            public void onFinish() {
                successDelete();
            }

            @Override
            public void onFail(final FTPFile iFTPFile) {
                super.onFail(iFTPFile);
                mHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        final AlertDialog.Builder lBuilder = new AlertDialog.Builder(mContextActivity);
                        lBuilder.setTitle("Delete error"); // TODO : strings
                        final View lDialogDeleteError = View.inflate(mContextActivity, R.layout.dialog_delete_error, null);
                        final CheckBox lCheckBox = lDialogDeleteError.findViewById(R.id.dialog_delete_error_checkbox);
                        final TextView lTextView = lDialogDeleteError.findViewById(R.id.dialog_delete_error_text);

                        lTextView.setText("Remember this choice : "); // TODO : strings
                        lBuilder.setView(lDialogDeleteError);
                        String lMessage =
                                "\nImpossible to delete the " +
                                        (iFTPFile.isDirectory() ? "directory \n" : "file \n") +
                                        iFTPFile.getName();
                        lBuilder.setMessage(lMessage);
                        lBuilder.setCancelable(false);
                        lBuilder.setPositiveButton("Ignore", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface iDialog, int iWhich) {
                                iDialog.dismiss();
                                mContextActivity.mDeletingErrorDialog = null;
                                if (mContextActivity.mDeletingInfoDialog != null)
                                    mContextActivity.mDeletingInfoDialog.show();
                                mContextActivity.mFTPServices.setDeletingByPassFailErrors(lCheckBox.isChecked());
                                mContextActivity.mFTPServices.resumeDeleting();
                            }
                        });

                        lBuilder.setNegativeButton("Stop", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface iDialog, int iWhich) {
                                iDialog.dismiss();
                                stopDeleting();
                            }
                        });
                        if (mContextActivity.mDeletingInfoDialog != null)
                            mContextActivity.mDeletingInfoDialog.hide();

                        mContextActivity.mDeletingErrorDialog = lBuilder.create();
                        mContextActivity.mDeletingErrorDialog.show();
                    }
                });
            }
        });
    }

    private void successDelete() {
        mHandler.sendEmptyMessage(NAVIGATION_ORDER_DISMISS_DIALOGS);
        mHandler.sendEmptyMessage(NAVIGATION_ORDER_REFRESH_DATA);
        mHandler.sendEmptyMessage(NAVIGATION_MESSAGE_DELETE_FINISHED);
    }

    private void stopDeleting() {
        mContextActivity.mCurrentAdapter.setSelectionMode(false);
        mContextActivity.mFTPServices.abortDeleting();
        mHandler.sendEmptyMessage(NAVIGATION_ORDER_DISMISS_DIALOGS);
        mHandler.sendEmptyMessage(NAVIGATION_ORDER_REFRESH_DATA);
        mHandler.sendEmptyMessage(NAVIGATION_MESSAGE_DELETE_FINISHED);
    }
}

File: app/src/main/java/com/vpulse/ftpnext/navigation/NavigationDelete.kt
package com.vpulse.ftpnext.navigation

import android.app.AlertDialog
import android.content.DialogInterface
import android.os.Handler
import android.view.View
import android.widget.CheckBox
import android.widget.ProgressBar
import android.widget.TextView
import com.vpulse.ftpnext.R
import com.vpulse.ftpnext.ftpservices.FTPServices
import org.apache.commons.net.ftp.FTPFile

class NavigationDelete(iContextActivity: NavigationActivity?, iHandler: Handler?) {

    private val mContextActivity: NavigationActivity? = iContextActivity
    private val mHandler: Handler? = iHandler

    fun onResume() {}
    fun createDialogDeleteSelection() {
        val lSelectedFiles: Array<FTPFile> = mContextActivity!!.mCurrentAdapter!!.selection
        if (lSelectedFiles.isEmpty()) {
            mContextActivity.createDialogError("Select something.").show()
        } else {
            mContextActivity.mDeletingInfoDialog =
                AlertDialog.Builder((mContextActivity)).setTitle("Deleting :") // TODO : Strings
                    .setMessage("Are you sure to delete the selection ? (" + lSelectedFiles.size + " files)")
                    .setPositiveButton("yes", object : DialogInterface.OnClickListener {
                        override fun onClick(iDialog: DialogInterface, iWhich: Int) {
                            iDialog.dismiss()
                            deleteFile(lSelectedFiles)
                        }
                    }).setNegativeButton("cancel", null).show()
        }
    }

    private fun deleteFile(iSelectedFiles: Array<FTPFile>?) {
        mHandler!!.sendEmptyMessage(NavigationActivity.NAVIGATION_ORDER_SELECTED_MODE_OFF)
        mContextActivity!!.mFTPServices!!.deleteFiles(iSelectedFiles!!,
            object : FTPServices.OnDeleteListener(mContextActivity.mFTPServices!!) {
                var mProgressDirectory: ProgressBar? = null
                var mProgressSubDirectory: ProgressBar? = null
                var mTextViewDirectory: TextView? = null
                var mTextViewSubDirectory: TextView? = null
                override fun onStartDelete() {
                    mHandler.post(Runnable {
                        val lBuilder: AlertDialog.Builder = AlertDialog.Builder(
                            (mContextActivity)
                        )
                        lBuilder.setTitle("Deleting") // TODO : strings
                        val lDialogDoubleProgress: View = View.inflate(
                            mContextActivity, R.layout.dialog_double_progress, null
                        )
                        mProgressDirectory =
                            lDialogDoubleProgress.findViewById<ProgressBar>(R.id.dialog_double_progress_progress_1)
                        mTextViewDirectory =
                            lDialogDoubleProgress.findViewById<TextView>(R.id.dialog_double_progress_text_1)
                        mProgressSubDirectory =
                            lDialogDoubleProgress.findViewById<ProgressBar>(R.id.dialog_double_progress_progress_2)
                        mTextViewSubDirectory =
                            lDialogDoubleProgress.findViewById<TextView>(R.id.dialog_double_progress_text_2)
                        lBuilder.setView(lDialogDoubleProgress)
                        lBuilder.setCancelable(false)
                        lBuilder.setNegativeButton("Cancel") { _, _ -> // TODO : Strings
                            stopDeleting()
                        }
                        mContextActivity.mDeletingInfoDialog = lBuilder.create()
                        mContextActivity.mDeletingInfoDialog!!.show()
                    })
                }

                override fun onProgressDirectory(iDirectoryProgress: Int,
                                                 iTotalDirectories: Int,
                                                 iDirectoryName: String?
                ) {
                    mHandler.post(Runnable {
                        mTextViewDirectory!!.text = iDirectoryName
                        mProgressDirectory!!.max = iTotalDirectories
                        mProgressDirectory!!.progress = iDirectoryProgress
                    })
                }

                override fun onProgressSubDirectory(iSubDirectoryProgress: Int,
                                                    iTotalSubDirectories: Int,
                                                    iSubDirectoryName: String?
                ) {
                    mHandler.post(Runnable {
                        mTextViewSubDirectory!!.text = iSubDirectoryName
                        mProgressSubDirectory!!.max = iTotalSubDirectories
                        mProgressSubDirectory!!.progress = iSubDirectoryProgress
                    })
                }

                override fun onRightAccessFail(iFTPFile: FTPFile) {
                    super.onRightAccessFail(iFTPFile)
                    mHandler.post(Runnable {
                        val lBuilder: AlertDialog.Builder = AlertDialog.Builder(
                            (mContextActivity)
                        )
                        lBuilder.setTitle("Permission error") // TODO : strings
                        val lDialogDeleteError: View = View.inflate(
                            mContextActivity, R.layout.dialog_delete_error, null
                        )
                        val lCheckBox: CheckBox =
                            lDialogDeleteError.findViewById<CheckBox>(R.id.dialog_delete_error_checkbox)
                        val lTextView: TextView =
                            lDialogDeleteError.findViewById<TextView>(R.id.dialog_delete_error_text)
                        lTextView.text = "Remember this choice : " // TODO : strings
                        lBuilder.setView(lDialogDeleteError)
                        val lMessage: String =
                            ("\nNot enough permissions to delete this " + (if (iFTPFile.isDirectory) "directory \n" else "file \n") + iFTPFile.name)
                        lBuilder.setMessage(lMessage)
                        lBuilder.setCancelable(false)
                        lBuilder.setPositiveButton(
                            "Ignore",
                            object : DialogInterface.OnClickListener {
                                override fun onClick(iDialog: DialogInterface, iWhich: Int
                                ) {
                                    iDialog.dismiss()
                                    mContextActivity.mFTPServices!!.setDeletingByPassRightErrors(
                                        lCheckBox.isChecked
                                    )
                                    mContextActivity.mFTPServices!!.resumeDeleting()
                                }
                            })
                        lBuilder.setNegativeButton(
                            "Stop",
                            object : DialogInterface.OnClickListener {
                                override fun onClick(iDialog: DialogInterface, iWhich: Int
                                ) {
                                    iDialog.dismiss()
                                    stopDeleting()
                                }
                            })
                        mContextActivity.mDeletingErrorDialog = lBuilder.create()
                        mContextActivity.mDeletingErrorDialog!!.show()
                    })
                }

                override fun onFinish() {
                    successDelete()
                }

                override fun onFail(iFTPFile: FTPFile?) {
                    super.onFail(iFTPFile)
                    mHandler.post(Runnable {
                        val lBuilder: AlertDialog.Builder = AlertDialog.Builder(
                            (mContextActivity)
                        )
                        lBuilder.setTitle("Delete error") // TODO : strings
                        val lDialogDeleteError: View = View.inflate(
                            mContextActivity, R.layout.dialog_delete_error, null
                        )
                        val lCheckBox: CheckBox =
                            lDialogDeleteError.findViewById<CheckBox>(R.id.dialog_delete_error_checkbox)
                        val lTextView: TextView =
                            lDialogDeleteError.findViewById<TextView>(R.id.dialog_delete_error_text)
                        lTextView.text = "Remember this choice : " // TODO : strings
                        lBuilder.setView(lDialogDeleteError)
                        val lMessage: String =
                            ("\nImpossible to delete the " + (if (iFTPFile!!.isDirectory) "directory \n" else "file \n") + iFTPFile.name)
                        lBuilder.setMessage(lMessage)
                        lBuilder.setCancelable(false)
                        lBuilder.setPositiveButton(
                            "Ignore",
                            object : DialogInterface.OnClickListener {
                                override fun onClick(iDialog: DialogInterface, iWhich: Int
                                ) {
                                    iDialog.dismiss()
                                    mContextActivity.mDeletingErrorDialog = null
                                    if (mContextActivity.mDeletingInfoDialog != null) mContextActivity.mDeletingInfoDialog!!.show()
                                    mContextActivity.mFTPServices!!.setDeletingByPassFailErrors(
                                        lCheckBox.isChecked
                                    )
                                    mContextActivity.mFTPServices!!.resumeDeleting()
                                }
                            })
                        lBuilder.setNegativeButton(
                            "Stop",
                            object : DialogInterface.OnClickListener {
                                override fun onClick(iDialog: DialogInterface, iWhich: Int
                                ) {
                                    iDialog.dismiss()
                                    stopDeleting()
                                }
                            })
                        if (mContextActivity.mDeletingInfoDialog != null) mContextActivity.mDeletingInfoDialog!!.hide()
                        mContextActivity.mDeletingErrorDialog = lBuilder.create()
                        mContextActivity.mDeletingErrorDialog!!.show()
                    })
                }
            })
    }

    private fun successDelete() {
        mHandler!!.sendEmptyMessage(NavigationActivity.Companion.NAVIGATION_ORDER_DISMISS_DIALOGS)
        mHandler.sendEmptyMessage(NavigationActivity.Companion.NAVIGATION_MESSAGE_DELETE_FINISHED)
    }

    private fun stopDeleting() {
        mContextActivity!!.mCurrentAdapter!!.setSelectionMode(false)
        mContextActivity.mFTPServices!!.abortDeleting()
        mHandler!!.sendEmptyMessage(NavigationActivity.Companion.NAVIGATION_ORDER_DISMISS_DIALOGS)
        mHandler.sendEmptyMessage(NavigationActivity.Companion.NAVIGATION_MESSAGE_DELETE_FINISHED)
    }

    companion object {
        private val TAG: String = "NAVIGATION DELETE"
    }
}